File: agent/workspace/boundary.py
"""File-level workspace boundary.

This module constrains paths supplied to tools.  It is not an OS sandbox:
commands still run as host processes with the workspace as their cwd.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any


def _inside(root: Path, path: Path) -> bool:
    try:
        path.relative_to(root)
        return True
    except ValueError:
        return False


def _reject_symlink_components(root: Path, path: Path) -> None:
    try:
        parts = path.relative_to(root).parts
    except ValueError as error:
        raise ValueError(f"path escapes workspace: {path}") from error
    current = root
    for part in parts:
        current /= part
        try:
            if current.is_symlink():
                raise ValueError(f"symlink path is not allowed: {path}")
        except OSError as error:
            raise ValueError(f"unable to inspect workspace path: {path}") from error
        if not current.exists():
            break


def workspace_path(root: str | Path, value: Any = ".", *, must_exist: bool = False) -> Path:
    """Resolve a tool path under ``root`` or the virtual ``/workspace`` root."""

    base = Path(root).resolve(strict=True)
    raw = str(value or ".")
    path = Path(raw)
    if path.is_absolute() and (path == Path("/workspace") or Path("/workspace") in path.parents):
        candidate = base / path.relative_to("/workspace")
    else:
        candidate = path if path.is_absolute() else base / path
    candidate = candidate.resolve(strict=False)
    if not _inside(base, candidate):
        raise ValueError(f"path escapes workspace: {value}")
    _reject_symlink_components(base, candidate)
    if must_exist and not candidate.exists():
        raise FileNotFoundError(candidate)
    return candidate

File: agent/workspace/boundary.ts
import { lstat, realpath } from "node:fs/promises";
import { dirname, isAbsolute, relative, resolve } from "node:path";

function inside(root: string, path: string): boolean {
  const distance = relative(root, path);
  return distance === "" || (!distance.startsWith("..") && !isAbsolute(distance));
}

async function rejectSymlinkComponents(root: string, path: string): Promise<void> {
  let current = root;
  const distance = relative(root, path);
  for (const part of distance ? distance.split(/[\\/]/) : []) {
    current = resolve(current, part);
    try {
      if ((await lstat(current)).isSymbolicLink()) throw new Error(`symlink path is not allowed: ${path}`);
    } catch (error: any) {
      if (error?.code === "ENOENT") break;
      throw error;
    }
  }
}

async function existingAncestor(path: string): Promise<string> {
  let current = path;
  while (true) {
    try { return await realpath(current); }
    catch (error: any) {
      if (error?.code !== "ENOENT") throw error;
      const parent = dirname(current);
      if (parent === current) throw error;
      current = parent;
    }
  }
}

/** Resolve a tool path under root or the virtual /workspace root. */
export async function workspacePath(root: string, value = ".", mustExist = false): Promise<string> {
  const base = await realpath(resolve(root));
  const raw = String(value || ".");
  const virtual = raw === "/workspace" || raw.startsWith("/workspace/");
  const lexical = virtual
    ? resolve(base, raw.slice("/workspace".length))
    : isAbsolute(raw) ? resolve(raw) : resolve(base, raw);
  if (!inside(base, lexical)) throw new Error("path escapes workspace");
  await rejectSymlinkComponents(base, lexical);
  const ancestor = await existingAncestor(lexical);
  if (!inside(base, ancestor)) throw new Error("path escapes workspace");
  try {
    const existing = await realpath(lexical);
    if (!inside(base, existing)) throw new Error("path escapes workspace");
    return existing;
  } catch (error: any) {
    if (error?.code !== "ENOENT") throw error;
    if (mustExist) throw error;
    return lexical;
  }
}
